### PR TITLE
Update keyvault to enable live testing in sovereign clouds for multiple services

### PR DESCRIPTION
**Pull request by v-xuto**
These changes enable keyvault to run live tests against Public, UsGov and China.
1. According to JS issue https://github.com/Azure/azure-sdk-for-js/issues/18267, there is the same error here . 
So skip the following tests:
- `test_key_client.py::KeyClientTests::test_key_crud_operations`, 
- `test_keys_async.py::KeyVaultKeyTest::test_key_crud_operations`,
- `test_samples_keys.py::TestExamplesKeyVault::test_example_key_crud_operations`,
- `test_samples_keys_async.py::TestExamplesKeyVault::test_example_key_crud_operations`
2. According to JS issue https://github.com/Azure/azure-sdk-for-js/issues/18142, there is the same error here. 
So skip the following tests:

- `test_key_client.py::KeyClientTests::test_key_rotation`,
- `test_key_client.py::KeyClientTests::test_key_rotation_policy`,
- `test_keys_async.py::KeyVaultKeyTest::test_key_rotation`,
- `test_keys_async.py::KeyVaultKeyTest::test_key_rotation_policy`

3. According to JS issue https://github.com/Azure/azure-sdk-for-js/issues/18414, there is the same error here. 
So add code `MatrixFilters:ArmTemplateParameters=^(?!.*enableHsm.*true)` in UsGov and China cloudconfig in tests.yml.

Pipeline results:
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1194909&view=results

@benbp , @jameszliao-msft , @lmazuel , @AlexGhiondea , @chlowell for notification.

### CODE REVIEW

**Comment by v-xuto** [comment]
@benbp Could you please review this PR?

File: sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py
Comment (chlowell) [inline sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py:189]: This test doesn't require a premium vault.

**Comment by benbp** [comment]
@v-xuto pipeline changes look good to me, but I defer to a language owner for reviewing the test code changes.

File: sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py
Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @chlowell I remove this code by your comments. But it failed in pipeline in China cloud. Error message `Hardware key operation not allowed on *** vault`. For more details please check [here](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1201262&view=logs&j=29c9017d-5931-5b87-96af-c798df6e68e8&t=97b61c76-f672-523b-e188-289102a5aa7d). 
Besides, I skip this test is based on the JS issue https://github.com/Azure/azure-sdk-for-js/issues/18267 and PR https://github.com/Azure/azure-sdk-for-js/pull/18350.
So, I think maybe we should skip this test when `KEYVAULT_SKU != "premium"`. Any ideas?

Comment (mccoyp) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: Hi @v-xuto, this actually is coming from a bug: when we [create a key](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py#L191) at the start of this test, we always create a hardware-protected key (`hardware_protected=True`) instead of doing so just for Managed HSM testing (`hardware_protected=is_hsm`). The same is true for the async version of this test, which is why they were failing for non-premium vaults.

Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @mccoyp I have remove this skip code in async version of this test. When do you plan to fix this test? What are your plans? We can only hold this PR first until you fix this test.

Comment (mccoyp) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @v-xuto I think we can fix the test as part of this PR, if that's alright, since we just need to change `hardware_protected=True` to `hardware_protected=is_hsm` [here](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py#L191) and [here](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/keyvault/azure-keyvault-keys/tests/test_keys_async.py#L186).

Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @mccoyp Thanks for your help. After I have updated by your comments, this issue about `test_key_crud_operations` has been resolved. But there is a new problem about `test_key_rotation_policy` in `public` cloud when I run it in a new pipeline. For more details please check [here](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1212260&view=logs&j=16f93b6d-ccf1-53e8-60ee-bc5b8717f42a&t=67369ee1-cdad-587c-15a2-c24641cb12fa).
Error message: `azure.core.exceptions.HttpResponseError: (Forbidden) The user, group or application 'appid=***;oid=***;iss=https://sts.windows.net/***/' does not have keys setrotationpolicy permission on key vault 't7d1580f70e294645;location=***'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287`
About this error, do you have any ideas?

Comment (mccoyp) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @v-xuto Sorry for the delay! I was out of office for a bit.

That error is coming from new permissions that are necessary for key release tests -- I just updated our tests in [this PR](https://github.com/Azure/azure-sdk-for-python/pull/21940) to adjust to the new requirements. If you rebase this PR with `main` then the error message should go away.

Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @mccoyp  Thank you for your PR, this PR fixed this error. But I found a new error `Failed to build cryptography.` in `ubuntu2004_pypy3`. I found that the pipeline that your PR runs also has this [problem](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1220316&view=results). Don't we need to care? If necessary, how do we solve it?

Comment (mccoyp) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @v-xuto Thank you for confirming! That `cryptography` issue is a known issue that popped up over the last week, yes. It should have been fixed with [this PR](https://github.com/Azure/azure-sdk-for-python/pull/21960), so re-running pipelines should resolve the issue.

Also, pipelines should pass for this PR if we update recordings for `test_key_client.py` and `test_keys_async.py`'s `test_key_crud_operations` test. Playback tests are [failing for now](https://dev.azure.com/azure-sdk/public/_build/results?buildId=1221643&view=ms.vss-test-web.build-test-results-tab&runId=26234412&resultId=100316&paneView=debug) because recordings for vault tests were always creating HSM keys. Would you be able to add new recordings after running that test live in sync and async, or would you like me to record and push to this PR?

Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @mccoyp  I rerun a new pipeline about this PR, it failed in `Deploy test resources`. For more details please check [here](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1224388&view=logs&j=16f93b6d-ccf1-53e8-60ee-bc5b8717f42a&t=6ce31d77-520f-5ca1-bdbb-af62fd4f0b50). Do you have any ideas? Besides, I will add recordings for `test_key_crud_operations` test later.

Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @benbp Do you have any suggestions about `deploy test resources failure`? For more details please check here.

Comment (mccoyp) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @v-xuto the `Deploy test resources` failure was caused by a bug that was fixed in [this PR](https://github.com/Azure/azure-sdk-tools/pull/2372), so a re-run of the pipeline should be able to deploy resources successfully

Comment (v-xuto) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @mccoyp I have add all recordings for `test_key_crud_operations` except for the `mhsm` key vault test. Because `mhsm` test not fail in ci pipeline and we run this live test locally hit the error `(Forbidden) The managed HSM resource is currently in Provisioned state. In this state you can only backup the current security domain or restore the security domain of another managed HSM resource`. Please review the recordings. Please let me know if there is any problem.

Comment (mccoyp) [reply sdk/keyvault/azure-keyvault-keys/tests/test_key_client.py]: @v-xuto Thank you for adding the recordings! I'll go ahead and approve, everything looks good to me. Thank you for all the work on this!!

**Review by mccoyp — APPROVED** [review]

**Comment by benbp** [comment]
@v-xuto assuming the tests pass is this ready to merge? It looks like there were some external dependency issues from yesterday's failure, so I'm re-running them.

**Comment by mikeharder** [comment]
@benbp: I cancelled the run of "python - keyvault - tests-weekly".  It was failing due to the DevOps cache task bug, and we're running low on agents due to provisioning issues.

**Comment by mccoyp** [comment]
@v-xuto I think you'll have to rebase this PR with the latest `main` to get the pipeline bug fix. @mikeharder or @benbp, please correct me if I'm wrong

**Comment by v-xuto** [comment]
@mccoyp I have rebased this PR with the latest `main`. Then I rerun a new weekly pipeline, but it also failed due to the DevOps cache task bug. About the ci pipeline error, it failed in the `Analyze dependencies` section.

**Comment by mccoyp** [comment]
@v-xuto Sorry about the failures! The `Analyze dependencies` error was just fixed earlier today with [this PR](https://github.com/Azure/azure-sdk-for-python/pull/22109), so another rebase (or copying the contents into this branch's `shared_requirements.txt`) should fix that issue.

As for the caching issue, that's something we're still working on. I made a temporary fix in the weekly tests pipeline, so re-running it should hopefully work. If it still fails, just let me know and we can provide an update when a more permanent fix is out.

**Comment by v-xuto** [comment]
@mccoyp Thanks for your help. Both of these issues have been resolved. I have rerun a new weekly pipeline, but it timed out in the `Deploy test resources` section in `ubuntu1804_ManagedHSM_39` about `public_azure_keyvault_keys` and  `public_azure_keyvault_administration`.
For more details please check [here](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1240898&view=logs&j=c810ea57-c63e-5f25-9a1d-6183954e69b1&t=39883043-b90d-5268-accd-a31693a7a2d).

**Comment by benbp** [comment]
@v-xuto we could potentially be running to conflicts with other HSM jobs in the same region depending on when it's ran. There's a strict quota for HSM resources per region within the same subscription, so it's possible that's the issue. I'm re-running to see if it works again.

**Comment by v-xuto** [comment]
@benbp This `quota for HSM resources` issue still exists when I run it in the last three days. Do we hold it temporarily or do you have any ideas to solve this problem?

**Comment by benbp** [comment]
@mccoyp it seems the HSM tests on python in general are failing/timing out. I wonder if the issue might be that HSM tests are running for both keys and administration (https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/keyvault/tests.yml#L20), meaning it will try to provision two HSM resources in the same region+subscription causing quota issues/delays?

**Comment by mccoyp** [comment]
@benbp True, I've been trying to figure out what's going on with HSM deployment for live pipelines. I don't think it's a quota issue; the service allowed five HSMs per region per subscription in preview, and there are usually no HSMs active in `westus` (each language team for KV uses a different region for pipeline testing). I was thinking that it could be an issue with `test-resources-post.ps1`, but after digging into it further, I noticed that HSM deployment stopped working last week without any code having been changed.

[This build](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1236930&view=results) from 12/8 had some service-related failures, but HSMs got deployed correctly. [Here](https://github.com/Azure/azure-sdk-for-python/blob/05a617a2766e35a0856b08ee97409ed549321d3f/sdk/keyvault/test-resources.json) is the `test-resources.json` file from the commit targeted in that build.

The next day, on 12/9, resource deployment failed in [this build](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1239528&view=results). The [test-resources.json](https://github.com/Azure/azure-sdk-for-python/blob/4db8247d519b48051eb9f88d1c2c44949b6a45ba/sdk/keyvault/test-resources.json) file from the build commit is identical to the one from the day before, and both commits have the same `test-resources-post.ps1` file as well. This makes me think that it must be a service issue, but [.NET pipelines](https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1247826&view=results) are having no issues with HSM deployment despite using the same [test-resources.json](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/keyvault/test-resources.json) (with minor differences like HSM region).

I do think it could be a good idea to deploy only one HSM for pipeline testing, and it might resolve this issue if simultaneous deployment is the problem.

**Comment by benbp** [comment]
@mccoyp perhaps we should try changing the HSM region and seeing if that works. Maybe there's a current issue in the westus region.

**Comment by mccoyp** [comment]
I created a PR [here](https://github.com/Azure/azure-sdk-for-python/pull/22179) to unblock HSM deployment

**Comment by benbp** [comment]
Thanks @v-xuto!

**Pull request merged**
